src/pycea/tl/permutation.py

```diff
@@ -0,0 +1,383 @@
+        pvalue_list = []
+        tree_list = []
+
+        data, is_array, is_square = get_keyed_obs_data(tdata, key)
+        if is_array or is_square and test == "t-test":
+            raise ValueError("t-test cannot be performed for vector valued keys.")
+        data = data.dropna()
+        index_set = set(data.index)
+        if not(is_array or is_square):
+            data = data[key]
+        for tree_id, t in trees.items():
+            for parent in nx.topological_sort(t):
+                children = list(t.successors(parent))
+
+                # don't do anything if not a split
+                if len(children) < 2:
+                    continue
+
+                # get leaves that are in the data
+                leaves_dict = {
+                    child: [u for u in get_leaves_from_node(t, child) if u in index_set]
+                    for child in children
+                }
+                for child, left_leaves in leaves_dict.items():
+                    # All other leaves except those from the current child
+                    right_leaves = [leaf for other_child, leaves in leaves_dict.items()
+                                    if other_child != child
+                                    for leaf in leaves]
+
+                    if len(left_leaves) > 0 and len(right_leaves) > 0:
+                        left_data = data.loc[left_leaves]
+                        right_data = data.loc[right_leaves]
+                    else:
+                        continue
+
+                    if copy and first_key:
+                        tree_list.append(tree_id)
+                        parent_list.append(parent)
+
+                    n_right = len(right_leaves)
+                    n_left = len(left_leaves)
+
+                    left_stat = aggregate_fn(left_data.to_numpy())
+                    right_stat = aggregate_fn(right_data.to_numpy())
+                    split_stat = float(metric_fn.pairwise(left_stat.reshape(1, -1), right_stat.reshape(1, -1)))
+
+                    nx.set_edge_attributes(t, {
+                        (parent, child): {f"{key_added}_value": left_stat}
+                    })
+
+                    if copy:
+                        group1_value_list.append(left_stat)
+                        group2_value_list.append(right_stat)
+                        if first_key:
+                            group1_list.append(child)
+
+                    if len(children) == 2:
+                        # handle special case in which there are exactly two children
+                        nx.set_edge_attributes(t, {
+                            (parent, children[1]): {f"{key_added}_value": right_stat}
+                        })
+
+                        if copy and first_key:
+                            group2_list.append(children[1])
+                    else:
+                        if copy and first_key:
+                            group2_list.append(", ".join([x for x in children if x != child]))
```

Comment:
Does not work for non string values.
```group2_list.append(", ".join(map(str, [x for x in children if x != child])))```

Reply:
I believe it should because I am calling map(str, ..)

```
child = 2
children = [1,2,3]
", ".join(map(str, [x for x in children if x != child]))
>>>
'1, 3'
```

Reply:
Unless we expect to have children names that are not even coercible to strings?